test/unit/architecture_test.rb

```diff
@@ -84,9 +92,45 @@
     context "parameters" do
       it_should_accept "name", ["--name=arch", "--new-name=arch2"]
       it_should_accept "id", ["--id=1", "--new-name=arch2"]
+      it_should_fail_with 'organization param', ['--organization-id=1']
+      it_should_fail_with 'location param', ['--location-id=1']
       # it_should_fail_with "no params", [] # TODO: temporarily disabled, parameters are checked in the id resolver
       # it_should_fail_with "name or id missing", ["--new-name=arch2"] # TODO: temporarily disabled, parameters are checked in the id resolver
     end
 
+    context 'ListCommandWithDefaults' do
+      let(:user) do
+        { 'default_organization' => { 'id' => 2 }, 'default_location' => { 'id' => 1 } }
+      end
+      let(:cmd) { HammerCLIForeman::Architecture::ListCommand.new('', ctx) }
+
+      before do
+        ResourceMocks.mock_action_call(:architectures, :index, [])
+        connection = api_connection
+        @api = APIExpectationsDecorator.new(connection.api)
+        @defaults_provider = HammerCLIForeman::Defaults.new(connection)
+      end
+
+      context 'parameters' do
+        it_should_accept 'no arguments'
+        it_should_accept_search_params
+        it_should_fail_with 'organization param', ['--organization-id=1']
+        it_should_fail_with 'location param', ['--location-id=1']
+      end
+
+      context 'output' do
+        it 'should run list command with defaults' do
+          @api.expects_search(:users, 'login=admin').returns(index_response([user]))
+          assert_equal 2, @defaults_provider.get_defaults(:organization_id)
+          cmd.resource.call(:index)
```

Comment:
Not sure if it actually tests anything.. I've tried to run `arch list` with foreman provider defaults and it does just one call to arch:index endpoint, no additional actions (for `host list` it does resolution for taxonomies which is correct). I guess the test should actually check that `api_expects_no_call` for getting defaults if the resource is not taxonomable (currently the test looks like it checks if architectures support defaults resolution for taxonomies, which is not right).

Also, such tests with api expectations should be moved among functional ones. The tests for params can and should be left among unit ones :)

Reply:
I see your point but I am not sure how to write these functional tests because I don't know how can I simulate in the tests that defaults were defined if no API call accrues when setting the defaults from hammer.

I hope I managed to explain my unclarity, If not let me know.

Reply:
@yifatmakias, yeap :) That was a little bit tricky, but a functional test for i.e. mail notifications will be:

```ruby
  it 'should run list command with defaults' do
    providers = { 'foreman' => HammerCLIForeman::Defaults.new(api_connection({}, '2.1')) }
    defaults = HammerCLI::Defaults.new(
      {
        organization_id: {
          provider: 'foreman'
        },
        location_id: {
          provider: 'foreman'
        }
      }
    )
    defaults.stubs(:write_to_file).returns(true)
    defaults.stubs(:providers).returns(providers)
    api_expects(:mail_notifications, :index, 'List mail notifications').returns(@mail_notifications)

    result = run_cmd(@cmd, { use_defaults: true, defaults: defaults })
    _(result.exit_code).must_equal HammerCLI::EX_OK
  end
```

This way we actually test that just one API call is being made and taxonomy options are ignored.

Also, it would be great to add at least one functional test for a resource that support default taxonomies, i.e. hosts:

```ruby
it 'should run list command with defaults' do
    providers = { 'foreman' => HammerCLIForeman::Defaults.new(api_connection({}, '2.1')) }
    defaults = HammerCLI::Defaults.new(
      {
        organization_id: {
          provider: 'foreman'
        },
        location_id: {
          provider: 'foreman'
        }
      }
    )
    defaults.stubs(:write_to_file).returns(true)
    defaults.stubs(:providers).returns(providers)
    api_expects(:users, :index, 'Find user').with_params(search: 'login=admin').returns(index_response([{ 'default_organization' => { 'id' => 2 } }]))
    api_expects(:users, :index, 'Find user').with_params(search: 'login=admin').returns(index_response([{ 'default_location' => { 'id' => 1 } }]))
    api_expects(:hosts, :index, 'List hosts').returns(index_response([{ 'id' => '42' }]))

    result = run_cmd(cmd, { use_defaults: true, defaults: defaults })
    _(result.exit_code).must_equal HammerCLI::EX_OK
  end
```

 This test shows that a command still works (badly optimized, but still) with taxonomy defaults.
 
 BTW, the tests could be reduced by extracting defaults/providers logic, but I'll probably leave it to you :rose:

Reply:
@ofedoren Thanks! I added these tests.